api/src/main/java/io/strimzi/api/kafka/model/KafkaUserAuthentication.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model;
+
+import io.strimzi.crdgenerator.annotations.Description;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import io.sundr.builder.annotations.Buildable;
+
+@Buildable(
+        editableEnabled = false,
+        generateBuilderPackage = true,
+        builderPackage = "io.strimzi.api.kafka.model"
+)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class KafkaUserAuthentication implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private Map<String, Object> additionalProperties;
+    private TlsClientAuthentication tls;
+
+    @Description("Enables TLS client authentication for given Kafka user")
+    @JsonInclude(value = JsonInclude.Include.NON_NULL)
+    public TlsClientAuthentication getTls() {
```

Comment:
If `TlsClientAuthentication` is supposed to be a Map of arbitrary things, just use the type `Map<String, Object>` instead of declaring a `TlsClientAuthentication` class with no properties and using it's additionalProperties. Alternatively `TlsClientAuthentication` should have getters and setters (or fields, which would make things less verbose).

Reply:
I don't know. I think in version (1) it will not be enything ... just empty object. Does setting it to Map allow us to detect whether it was set or whether it was omitted?

Reply:
I'm still not completely sure I understand what were you asking for here.

Reply:
I suppose I'm trying to understand how this should look once we support more than just TLS as an auth mechanism (and possibly auth mechanisms which need some configuration). 

What you have currently is this:

```
spec:
  authentication:
    tls: {}
```

Which is a bit odd if `tls` doesn't – and never will have – any properties. Also, once we support other mechanisms, would we allow:


```
spec:
  authentication:
    tls: {}
    scramSha: {}
```

Or is the mechanism either `tls` or `scramSha`? While I suppose it sort-of makes sense that you could use both it's definitely a strange thing to do, and I can't think of any use cases for it. In which case it would probably be better to model it as:


```
spec:
  authentication:
    type: tls
```

and 


```
spec:
  authentication:
    type: scramSha
    whatever: ScramShaSpecificProperty
```

Reply:
So I can actually imagine the `tls` to have properties. I can for example imagine something like this:

```
spec:
  authentication:
    tls:
      subject:
        org: MyOrg
        orgUnit: MyTeam
```

Which would allow the user to configure the details of the certificate subject (CN would be the user name == name of the KafkaUSer CR).

Reply:
The discussion between:

```
spec:
  authentication:
    tls: {}
    scramSha: {}
```

and 

```
spec:
  authentication:
    type: tls
```

That is basically about being able to specify more auth methods for one user or not. I think we discussed this in your Gist and I actually thought your idea was to support multiple. We can change it to use the type which would support only one mechanism per user.